receiver/prometheusreceiver/internal/transaction.go

```diff
@@ -166,7 +166,7 @@ func (tr *transaction) initTransaction(ls labels.Labels) error {
 		tr.instance = instance
 	}
 	tr.node, tr.resource = createNodeAndResource(job, instance, mc.SharedLabels().Get(model.SchemeLabel))
-	tr.metricBuilder = newMetricBuilder(mc, tr.useStartTimeMetric, tr.startTimeMetricRegex, tr.logger, tr.stalenessStore)
+	tr.metricBuilder = newMetricBuilder(mc, tr.useStartTimeMetric, tr.startTimeMetricRegex, tr.logger, tr.stalenessStore, tr.startTimeMs)
```

Comment:
It looks like an Appender is created when the config is loaded: https://github.com/prometheus/prometheus/blob/5cc30b716e23c98863093c0ed39315c1fe7c103e/scrape/manager.go#L257.  That would mean that this start time is the start time of the first metric that the collector sees, which is approximately the start time of the collector.  That seems like it would be incorrect in many cases.  Is there something i'm missing?

Reply:
Nope, for every Scrape,  a new storage.Appender is requested, not when the collector is started. This means that
even a new transaction is created per scrape and at the end either .Commit() or .Rollback() is invoked. This isn't a problem.

Reply:
Got it, thanks.  Is the storage.Appender created after the scrape finishes, or before the scrape is started?  We just need to make sure the start time isn't before the metric's timestamp

Reply:
Another option for the moment is to leave the starttime unset (0) which is conforming with our specs, then fix this the correct timestamp later. This is to unblock the progress.

Reply:
>Got it, thanks. Is the storage.Appender created after the scrape finishes, or before the scrape is started? We just need to make sure the start time isn't before the metric's timestamp

@dashpole we infer that startTime for the very first time that's passed into (*transaction).Append so it will ALWAYS be the first timestamp. initTransaction is lazily loaded and .startTimeMs will ALWAYS be set after the first .Append.
https://github.com/open-telemetry/opentelemetry-collector/blob/f804c9c4431f9c98d15e65c257742f2c3daa3e32/receiver/prometheusreceiver/internal/transaction.go#L116-L118
https://github.com/open-telemetry/opentelemetry-collector/blob/f804c9c4431f9c98d15e65c257742f2c3daa3e32/receiver/prometheusreceiver/internal/transaction.go#L136-L140
@dashpole to answer your question about Prometheus (but it is irrelevant to this code) the storage.Appender is created before the scrape begins per https://github.com/prometheus/prometheus/blob/1a430e5f894d6575478bcef52d25bfbe7134947e/scrape/scrape.go#L1182-L1237 or visually
![Screen Shot 2021-07-29 at 2 19 49 PM](https://user-images.githubusercontent.com/4898263/127567345-946cfa37-02e0-439c-9c6c-edbaa08dcc39.png)
![Screen Shot 2021-07-29 at 2 19 58 PM](https://user-images.githubusercontent.com/4898263/127567365-717e2caa-95fa-468a-a5e0-f0f610937037.png)


>Another option for the moment is to leave the starttime unset (0) which is conforming with our specs, then fix this the correct timestamp later. This is to unblock the progress.

@bogdandrutu thanks for raising this, I didn't know about using 0 conforming with specs. No, we are already good, the intervalStartTime is inferred from the very first start time when (*transaction).Append so we are all set.